share/man/man4/lpt.4
.\" $NetBSD$
.\"
.\"
.\" Copyright (c) 1993 Christopher G. Demetriou
.\" Copyright (c) 1994 Geoffrey M. Rehmet
.\" Copyright (c) 1999 Nicolas Souchu
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" $FreeBSD: src/share/man/man4/lpt.4,v 1.9.2.2 2001/08/17 13:08:38 ru Exp $
.\"
.Dd February 3, 2004
.Dt LPT 4
.Os
.Sh NAME
.Nm lpt
.Nd generic printer device driver
.Sh SYNOPSIS
.Cd "lpt* at ppbus?"
.Cd options LPT_DEBUG
.Cd options LPT_VERBOSE
.Sh DESCRIPTION
The
.Nm
driver is the port of the lpt driver to the
.Xr ppbus 4
system.
.Pp
The purpose of this driver is to allow parallel port sharing with
other parallel devices.
See
.Xr ppbus 4
for more information about the parallel port bus system.
.Pp
The parallel port bus is reserved by
.Nm
when the printer device is opened and released when the device is
closed.
.Pp
Ports can be configured to use interrupts, DMA, IEEE negotiations, and IEEE
compliant transfer modes by using the
.Xr lptctl 8
command, with modes depending on the hardware available.
.Sh FILES
.Bl -tag -width "/dev/lpt0xxxx" -compact
.It Pa /dev/lpt?
parallel port device
.It Pa /dev/lpt?ctl
parallel port control device
.El
.Sh SEE ALSO
.Xr atppc 4 ,
.Xr ppbus 4 ,
.Xr lptctl 8 ,
.Xr mknod 8
.Sh HISTORY
This driver is derived from the
.Fx
ppbus and also from the historic
.Nx
drivers.
.Pp
The
.Nm
driver used to assign special meaning for some bits of device minor.
This was dropped in ppbus-based
.Nm
in favour of configuration via
.Xr lptctl 8 .
.Sh AUTHORS
This manual page was based on the
.Fx
.Nm lpt
manual page.
The information has been updated for the
.Nx
port by Gary Thorpe.


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision$>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 6.0 to NetBSD 7.0:
	window(1): Removed from the base system in favor of tmux(1).  Use the
		pkgsrc/misc/window package from now on.  [jmmv 20120215]
	pkg_add(1): imported 20120221 version from pkgsrc [wiz 20120221]
	file(1): upgraded to 5.11 [christos 20120222]
	evbmips: Add support for Lemote Yeeloong Notebook. [nonaka 20120302]
	dbsym(8): Add option -p [bsh 20120319]
	dhcpcd(8): Import dhcpcd-5.5.5 [roy 20120323]
	ipf: upgrade to 5.1.1 with new build framework, no module and
		test support yet [christos 20120323]
	pcc: upgrade to 1.1.0.DEVEL 20120325 [plunky 20120325]
	openresolv(8): Import openresolv-3.5.1 [roy 20120327]
	dhcpcd(8): Import dhcpcd-5.5.6 [roy 20120328]
	apple: Recognize the apple disk partitioning scheme, and auto-discover
		wedges. Fix iPod disk mounting, by adding quirk.
		[christos 20120306]
	fstab: Add NAME=wedgename for fs_spec and routines to parse it. Now
		wedges can be used in /etc/fstab, even in dynamic environments
		where disks come and go and wedges get renumbered.
		[christos 20120306]
	pdisk(8): Make it work for 64 bit hosts, and make it MI.
		[christos 20120306]
	amiga: Add clockport(4) layer and a driver for A1200 on-board clockport
		a1k2cp(4). [rkujawa 20120417]
	kernel: add mpii(4), a driver for LSI Logic Fusion-MPT Message Passing
		Interface II SAS controllers.  [bouyer 20120419]
	kernel: Add support for 256 CPUs on amd64. [rmind 20120420]
	OpenSSH: Imported 6.0 [christos 20120501]
	bzip2: Imported 1.0.6 [wiz 20120507]
	amiga: Rework X-Surf driver as xsurf(4). Add support for clockport(4)
		on top of xsurf(4). [rkujawa 20120515]
	backtrace(3): Add <execinfo.h> and the backtrace family of functions
		[christos 20120526]
	kernel: Add run(4), a driver for Ralink Technology
		RT2700U/RT2800U/RT3000U USB IEEE 802.11a/b/g/n wireless
		network devices, ported from OpenBSD. [nonaka 20120530]
	bind: update to 9.9.1-P1 to fix the recent rdata exploit.
		[christos 20120604]
	empb(4): Add ELBOX Mediator PCI 1200 driver. [rkujawa 20120604]
	kernel: Add a new RLIMIT_NTHR limit to limit the number of processes
		per user and the corresponding kern.maxlwp sysctl. Expose
		kern.uidinfo.* via sysctl. [christos 20120608]
	usb(9): Added support for SMP to the USB subsytem.  By Jared D.
		McNeill and myself.  [mrg 20120609]
	kernel: {send,recv}mmsg support [christos 20120621]
	kernel: Add sysctls to avoid ipv6 DoS attacks (from OpenBSD):
		net.inet6.ip6.neighborgcthresh = 2048
		net.inet6.ip6.maxifprefixes = 16
		net.inet6.ip6.maxifdefrouters = 16
		net.inet6.ip6.maxdynroutes = 4096
		[christos 20120622]
	ata(4): added support for sata Port MultiPliers (PMP)
		[bouyer 20120702]
	kernel, libc: Add MurmurHash2 function. [rmind 20120708]
	atf(7): Import 0.16.  [jmmv 20120711]
	dhcpcd(8): Import dhcpcd-5.6.1 [roy 20120712]
	kernel: Add support for sensors to provide entropy to rnd(4)
		[pgoyette 20120715]
	tdvfb(4): Add 3Dfx Voodoo2 driver. [rkujawa 20120719]
	tdvfb(4): Add 3Dfx Voodoo Graphics (aka Voodoo1) support.
		[rkujawa 20120720]
	kernel: Change mii_statchg to take "struct ifnet *" instead of
		"device_t". [matt 20120722]
	pr(1): Add support for POSIX -f and -p options. [ginsbach 20120724]
	powerpc: Add support for FPU emulation on BookE. [matt 20120722]
	OpenSSL: Imported 1.0.1c [christos 20120726]
	kernel: safepri is dead; replaced by macro IPL_SAFEPRI. [matt 20120727]
	kernel: Add malo(4), a driver for Marvell Libertas IEEE 802.11b/g
		wireless network devices, ported from OpenBSD.
		[degroote 20120730]
	kernel, libc: Add modctl(MODCTL_EXISTS, ...) to determine if modules
		can be loaded. [jnemeth 20120803]
	modstat(8): Add -A, -a, and -e options for testing module loadability.
		[jnemeth 20120803]
	libc: Update to tzcode2012e. [christos 20120809]
	arm: Add support for BEAGLEBOARDXM (TI DM37XX cpus) boards.
		[matt 20120822]
	mfi(8): Added tagged queuing support.  [bouyer 20120823]
	mfi(8): Added support for LSI SAS2208-based controllers.
		[bouyer 20120823]
	mkdep(1): Add -P option.  [jmmv 20120826]
	share/mk: Add placeholder test programs (with some very basic test
		cases within) for common bsd.*.mk modules.  [jmmv 20120826]
	arm: Add Cortex-A9 support (a9tmr, a9wdt, armgic, armperiph).  Add
		initial MP support (can boot 2nd cpu but dies quickly).  Add
		L1 SuperSection support for ARMv6/ARMv7.  Add support for 
		BCM5301X SoC.  Add initial support for OMAP4430 and partial
		support for Freescale i.MX6.  Add new common boot/kvminit
		code.  [matt 20120831]
	dhcpcd(8): Import dhcpcd-5.6.2 [roy 20120903]

sys/dev/ppbus/lpt.c
/* $NetBSD$ */

/*
 * Copyright (c) 1990 William F. Jolitz, TeleMuse
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This software is a component of "386BSD" developed by
 *	William F. Jolitz, TeleMuse.
 * 4. Neither the name of the developer nor the name "386BSD"
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS A COMPONENT OF 386BSD DEVELOPED BY WILLIAM F. JOLITZ
 * AND IS INTENDED FOR RESEARCH AND EDUCATIONAL PURPOSES ONLY. THIS
 * SOFTWARE SHOULD NOT BE CONSIDERED TO BE A COMMERCIAL PRODUCT.
 * THE DEVELOPER URGES THAT USERS WHO REQUIRE A COMMERCIAL PRODUCT
 * NOT MAKE USE OF THIS WORK.
 *
 * FOR USERS WHO WISH TO UNDERSTAND THE 386BSD SYSTEM DEVELOPED
 * BY WILLIAM F. JOLITZ, WE RECOMMEND THE USER STUDY WRITTEN
 * REFERENCES SUCH AS THE  "PORTING UNIX TO THE 386" SERIES
 * (BEGINNING JANUARY 1991 "DR. DOBBS JOURNAL", USA AND BEGINNING
 * JUNE 1991 "UNIX MAGAZIN", GERMANY) BY WILLIAM F. JOLITZ AND
 * LYNNE GREER JOLITZ, AS WELL AS OTHER BOOKS ON UNIX AND THE
 * ON-LINE 386BSD USER MANUAL BEFORE USE. A BOOK DISCUSSING THE INTERNALS
 * OF 386BSD ENTITLED "386BSD FROM THE INSIDE OUT" WILL BE AVAILABLE LATE 1992.
 *
 * THIS SOFTWARE IS PROVIDED BY THE DEVELOPER ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE DEVELOPER BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *	from: unknown origin, 386BSD 0.1
 *	From Id: lpt.c,v 1.55.2.1 1996/11/12 09:08:38 phk Exp
 *	From Id: nlpt.c,v 1.14 1999/02/08 13:55:43 des Exp
 * FreeBSD: src/sys/dev/ppbus/lpt.c,v 1.15.2.3 2000/07/07 00:30:40 obrien Exp
 */

/*
 * Device Driver for AT parallel printer port
 * Written by William Jolitz 12/18/90
 */

/*
 * Updated for ppbus by Nicolas Souchu
 * [Mon Jul 28 1997]
 */

#include <sys/cdefs.h>
__KERNEL_RCSID(0, "$NetBSD$");

#include "opt_ppbus_lpt.h"

#include <sys/param.h>
#include <sys/systm.h>
#include <sys/conf.h>
#include <sys/kernel.h>
#include <sys/proc.h>
#include <sys/malloc.h>
#include <sys/file.h>
#include <sys/uio.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/syslog.h>

#include <sys/bus.h>

#include <dev/ppbus/ppbus_1284.h>
#include <dev/ppbus/ppbus_base.h>
#include <dev/ppbus/ppbus_io.h>
#include <dev/ppbus/ppbus_msq.h>
#include <dev/ppbus/ppbus_var.h>

#include <dev/ppbus/lptvar.h>
#include <dev/ppbus/lptreg.h>
#include <dev/ppbus/lptio.h>

/* Autoconf functions */
static int lpt_probe(device_t, cfdata_t, void *);
static void lpt_attach(device_t, device_t, void *);
static int lpt_detach(device_t, int);

/* Autoconf structure */
CFATTACH_DECL_NEW(lpt_ppbus, sizeof(struct lpt_softc), lpt_probe, lpt_attach,
	lpt_detach, NULL);

extern struct cfdriver lpt_cd;

dev_type_open(lptopen);
dev_type_close(lptclose);
dev_type_read(lptread);
dev_type_write(lptwrite);
dev_type_ioctl(lptioctl);

const struct cdevsw lpt_cdevsw = {
        lptopen, lptclose, lptread, lptwrite, lptioctl,
	nostop, notty, nopoll, nommap, nokqfilter, D_OTHER
};


/* Function prototypes */
static int lpt_detect(device_t);
static int lpt_request_ppbus(struct lpt_softc *, int);
static int lpt_release_ppbus(struct lpt_softc *, int);
static int lpt_logstatus(const device_t, const unsigned char);

/*
 * lpt_probe()
 */
static int
lpt_probe(device_t parent, cfdata_t match, void *aux)
{
	/* Test ppbus's capability */
	return lpt_detect(parent);
}

static void
lpt_attach(device_t parent, device_t self, void *aux)
{
	struct lpt_softc * sc = device_private(self);
	struct ppbus_device_softc * ppbdev = &(sc->ppbus_dev);
	struct ppbus_attach_args * args = aux;
	char buf[64];
	int error;

	ppbdev->sc_dev = self;

	error = lpt_request_ppbus(sc, 0);
	if(error) {
		printf("%s(%s): error (%d) requesting bus(%s). Device not "
			"properly attached.\n", __func__, device_xname(self),
			error, device_xname(parent));
		return;
	}

	/* Record capabilities */
	ppbdev->capabilities = args->capabilities;

	/* Allocate memory buffers */
	if(ppbdev->capabilities & PPBUS_HAS_DMA) {
		if(ppbus_dma_malloc(parent, &(sc->sc_inbuf),
			&(sc->sc_in_baddr), BUFSIZE)) {

			printf(" : cannot allocate input DMA buffer. Device "
				"not properly attached!\n");
			return;
		}
		if(ppbus_dma_malloc(parent, &(sc->sc_outbuf),
			&(sc->sc_out_baddr), BUFSIZE)) {

			ppbus_dma_free(parent, &(sc->sc_inbuf),
				&(sc->sc_in_baddr), BUFSIZE);
			printf(" : cannot allocate output DMA buffer. Device "
				"not properly attached!\n");
			return;
		}
	} else {
		sc->sc_inbuf = malloc(BUFSIZE, M_DEVBUF, M_WAITOK);
		sc->sc_outbuf = malloc(BUFSIZE, M_DEVBUF, M_WAITOK);
	}

	/* Print out mode */
        ppbdev->ctx.mode = ppbus_get_mode(parent);
	snprintb(buf, sizeof(buf),
	    "\20\1COMPATIBLE\2NIBBLE\3PS2\4EPP\5ECP\6FAST_CENTR",
	    ppbdev->ctx.mode);
	printf(": port mode = %s\n", buf);

	/* Initialize the device on open by default */
	sc->sc_flags = LPT_PRIME;

	lpt_release_ppbus(sc, 0);
}

static int
lpt_detach(device_t self, int flags)
{
	struct lpt_softc * lpt = device_private(self);
	struct ppbus_device_softc * ppbdev = (struct ppbus_device_softc *) lpt;
	int err;

	if(lpt->sc_state & HAVEBUS) {
		err = lpt_release_ppbus(lpt, 0);
		if(err) {
			printf("%s error (%d) while releasing bus",
				device_xname(self), err);
			if(flags & DETACH_FORCE) {
				printf(", continuing (DETACH_FORCE)!\n");
			}
			else {
				printf(", terminating!\n");
				return 0;
			}
		}
		lpt->sc_state &= ~HAVEBUS;
	}

	ppbdev->ctx.valid = 0;

	/* Free memory buffers */
	if(ppbdev->capabilities & PPBUS_HAS_DMA) {
		ppbus_dma_free(device_parent(self), &(lpt->sc_inbuf),
			&(lpt->sc_in_baddr), BUFSIZE);
		ppbus_dma_free(device_parent(self), &(lpt->sc_outbuf),
			&(lpt->sc_out_baddr), BUFSIZE);
	} else {
		free(lpt->sc_inbuf, M_DEVBUF);
		free(lpt->sc_outbuf, M_DEVBUF);
	}

	return 1;
}

/* Grab bus for lpt device */
static int
lpt_request_ppbus(struct lpt_softc * lpt, int how)
{
	device_t dev = lpt->ppbus_dev.sc_dev;
	int error;

	error = ppbus_request_bus(device_parent(dev), dev, how, (hz));
	if (!(error)) {
		lpt->sc_state |= HAVEBUS;
	}
	else {
		LPT_DPRINTF(("%s(%s): error %d requesting bus.\n", __func__,
			device_xname(dev), error));
	}

	return error;
}

/* Release ppbus to enable other devices to use it. */
static int
lpt_release_ppbus(struct lpt_softc * lpt, int how)
{
	device_t dev = lpt->ppbus_dev.sc_dev;
	int error;

	if(lpt->sc_state & HAVEBUS) {
		error = ppbus_release_bus(device_parent(dev), dev, how, (hz));
		if(!(error))
			lpt->sc_state &= ~HAVEBUS;
		else {
			LPT_DPRINTF(("%s(%s): error releasing bus.\n", __func__,
				device_xname(dev)));
		}
	}
	else {
		error = EINVAL;
		LPT_DPRINTF(("%s(%s): device does not own bus.\n", __func__,
			device_xname(dev)));
	}

	return error;
}


/*
 * Probe simplified by replacing multiple loops with a hardcoded
 * test pattern - 1999/02/08 des@freebsd.org
 *
 * New lpt port probe Geoff Rehmet - Rhodes University - 14/2/94
 * Based partially on Rod Grimes' printer probe
 *
 * Logic:
 *	1) If no port address was given, use the bios detected ports
 *	   and autodetect what ports the printers are on.
 *	2) Otherwise, probe the data port at the address given,
 *	   using the method in Rod Grimes' port probe.
 *	   (Much code ripped off directly from Rod's probe.)
 *
 * Comments from Rod's probe:
 * Logic:
 *	1) You should be able to write to and read back the same value
 *	   to the data port.  Do an alternating zeros, alternating ones,
 *	   walking zero, and walking one test to check for stuck bits.
 *
 *	2) You should be able to write to and read back the same value
 *	   to the control port lower 5 bits, the upper 3 bits are reserved
 *	   per the IBM PC technical reference manauls and different boards
 *	   do different things with them.  Do an alternating zeros, alternating
 *	   ones, walking zero, and walking one test to check for stuck bits.
 *
 *	   Some printers drag the strobe line down when the are powered off
 * 	   so this bit has been masked out of the control port test.
 *
 *	   XXX Some printers may not like a fast pulse on init or strobe, I
 *	   don't know at this point, if that becomes a problem these bits
 *	   should be turned off in the mask byte for the control port test.
 *
 *	   We are finally left with a mask of 0x14, due to some printers
 *	   being adamant about holding other bits high ........
 *
 *	   Before probing the control port, we write a 0 to the data port -
 *	   If not, some printers chuck out garbage when the strobe line
 *	   gets toggled.
 *
 *	3) Set the data and control ports to a value of 0
 *
 *	This probe routine has been tested on Epson Lx-800, HP LJ3P,
 *	Epson FX-1170 and C.Itoh 8510RM
 *	printers.
 *	Quick exit on fail added.
 */
static int
lpt_detect(device_t dev)
{
	static const u_char testbyte[18] = {
		0x55,			/* alternating zeros */
		0xaa,			/* alternating ones */
		0xfe, 0xfd, 0xfb, 0xf7,
		0xef, 0xdf, 0xbf, 0x7f,	/* walking zero */
		0x01, 0x02, 0x04, 0x08,
		0x10, 0x20, 0x40, 0x80	/* walking one */
	};
	int i, status;
	u_char dtr, ctr, str, var;

	/* Save register contents */
	dtr = ppbus_rdtr(dev);
	ctr = ppbus_rctr(dev);
	str = ppbus_rstr(dev);

	status = 1;				/* assume success */

	/* Test data port */
	for(i = 0; i < 18; i++) {
		ppbus_wdtr(dev, testbyte[i]);
		if((var = ppbus_rdtr(dev)) != testbyte[i]) {
			status = 0;
			LPT_DPRINTF(("%s(%s): byte value %x cannot be written "
				"and read from data port (got %x instead).\n",
				__func__, device_xname(dev), testbyte[i], var));
			goto end;
		}
	}

	/* Test control port */
	ppbus_wdtr(dev, 0);
	for(i = 0; i < 18; i++) {
		ppbus_wctr(dev, (testbyte[i] & 0x14));
		if(((var = ppbus_rctr(dev)) & 0x14) != (testbyte[i] & 0x14)) {
			status = 0;
			LPT_DPRINTF(("%s(%s): byte value %x (unmasked value "
				"%x) cannot be written and read from control "
				"port (got %x instead).\n", __func__,
				device_xname(dev), (testbyte[i] & 0x14),
				testbyte[i], (var & 0x14)));
			break;
		}
	}

end:
	/* Restore contents of registers */
	ppbus_wdtr(dev, dtr);
	ppbus_wctr(dev, ctr);
	ppbus_wstr(dev, str);

	return status;
}

/* Log status of status register for printer port */
static int
lpt_logstatus(const device_t dev, const unsigned char status)
{
	int err;

	err = EIO;
	if(!(status & LPS_SEL)) {
		log(LOG_ERR, "%s: offline.", device_xname(dev));
	}
	else if(!(status & LPS_NBSY)) {
		log(LOG_ERR, "%s: busy.", device_xname(dev));
	}
	else if(status & LPS_OUT) {
		log(LOG_ERR, "%s: out of paper.", device_xname(dev));
		err = EAGAIN;
	}
	else if(!(status & LPS_NERR)) {
		log(LOG_ERR, "%s: output error.", device_xname(dev));
	}
	else {
		log(LOG_ERR, "%s: no error indication.", device_xname(dev));
		err = 0;
	}

	return err;
}

/*
 * lptopen -- reset the printer, then wait until it's selected and not busy.
 */
int
lptopen(dev_t dev_id, int flags, int fmt, struct lwp *l)
{
	int trys, err;
	u_int8_t status;
	device_t dev;
	struct lpt_softc * lpt;
	struct ppbus_device_softc * ppbus_dev;
	device_t ppbus;

	dev = device_lookup(&lpt_cd, LPTUNIT(dev_id));
	if(!dev) {
		LPT_DPRINTF(("%s(): device not configured.\n", __func__));
		return ENXIO;
	}

	lpt = device_private(dev);

	ppbus = device_parent(dev);
	ppbus_dev = &(lpt->ppbus_dev);

	/* Request the ppbus */
	err = lpt_request_ppbus(lpt, PPBUS_WAIT|PPBUS_INTR);
	if(err) {
		LPT_DPRINTF(("%s(%s): error (%d) while requesting bus.\n",
			__func__, device_xname(dev), err));
		return (err);
	}

	/* Update bus mode */
	ppbus_dev->ctx.mode = ppbus_get_mode(ppbus);

	/* init printer */
	if ((lpt->sc_flags & LPT_PRIME) && !LPTCTL(dev_id)) {
		LPT_VPRINTF(("%s(%s): initializing printer.\n", __func__,
			device_xname(dev)));
		lpt->sc_state |= LPTINIT;
		ppbus_wctr(ppbus, LPC_SEL | LPC_NINIT);

		/* wait till ready (printer running diagnostics) */
		for(trys = 0, status = ppbus_rstr(ppbus); (status & RDY_MASK)
			!= LP_READY; trys += LPT_STEP, status =
			ppbus_rstr(ppbus)) {

			/* Time up waiting for the printer */
			if(trys >= LPT_TIMEOUT)
				break;
			/* wait LPT_STEP ticks, give up if we get a signal */
			else {
				err = tsleep((void *)lpt, LPPRI|PCATCH,
					"lptinit", LPT_STEP);
				if((err) && (err != EWOULDBLOCK)) {
					lpt->sc_state &= ~LPTINIT;
					LPT_DPRINTF(("%s(%s): interrupted "
					"during initialization.\n", __func__,
					device_xname(dev)));
					lpt_release_ppbus(lpt, PPBUS_WAIT);
					return (err);
				}
			}
		}

		lpt->sc_state &= ~LPTINIT;
		if(trys >= LPT_TIMEOUT) {
			LPT_DPRINTF(("%s(%s): timed out while initializing "
				"printer. [status %x]\n", __func__,
				device_xname(dev), status));
			err = lpt_logstatus(dev, status);
			lpt_release_ppbus(lpt, PPBUS_WAIT);
			return (err);
		}
		else {
			LPT_VPRINTF(("%s(%s): printer ready.\n", __func__,
				device_xname(dev)));
		}
	}

	/* Set autolinefeed if requested */
	if (lpt->sc_flags & LPT_AUTOLF)
		ppbus_wctr(ppbus, LPC_AUTOL);
	else
		ppbus_wctr(ppbus, 0);

	/* ready now */
	lpt->sc_state |= OPEN;

	return 0;
}

/*
 * lptclose -- close the device, free the local line buffer.
 *
 * Check for interrupted write call added.
 */
int
lptclose(dev_t dev_id, int flags, int fmt, struct lwp *l)
{
	device_t dev = device_lookup(&lpt_cd, LPTUNIT(dev_id));
	struct lpt_softc *sc = device_private(dev);
	int err;

	err = lpt_release_ppbus(sc, PPBUS_WAIT|PPBUS_INTR);
	if(err) {
		LPT_DPRINTF(("%s(%s): error (%d) while releasing ppbus.\n",
			__func__, device_xname(dev), err));
	}

	sc->sc_state = 0;

	return err;
}

/*
 * lptread --retrieve printer status in IEEE1284 NIBBLE mode
 */
int
lptread(dev_t dev_id, struct uio *uio, int ioflag)
{
	size_t len = 0;
	int error = 0;
	device_t dev = device_lookup(&lpt_cd, LPTUNIT(dev_id));
	struct lpt_softc *sc = device_private(dev);

	if(!(sc->sc_state & HAVEBUS)) {
		LPT_DPRINTF(("%s(%s): attempt to read using device which does "
			"not own the bus(%s).\n", __func__, device_xname(dev),
			device_xname(device_parent(dev))));
		return (ENODEV);
	}

	sc->sc_state &= ~INTERRUPTED;
	while (uio->uio_resid) {
		error = ppbus_read(device_parent(dev), sc->sc_outbuf,
			min(BUFSIZE, uio->uio_resid), 0, &len);

		/* If error or no more data, stop */
		if (error) {
			if (error != EWOULDBLOCK)
				sc->sc_state |= INTERRUPTED;
			break;
		}
		if (len == 0)
			break;

		if ((error = uiomove(sc->sc_outbuf, len, uio)))
			break;
	}

	return error;
}

/*
 * lptwrite --copy a line from user space to a local buffer, then call
 * putc to get the chars moved to the output queue.
 *
 * Flagging of interrupted write added.
 */
int
lptwrite(dev_t dev_id, struct uio * uio, int ioflag)
{
	int error=0;
	size_t n, cnt;
	device_t dev = device_lookup(&lpt_cd, LPTUNIT(dev_id));
	struct lpt_softc * sc = device_private(dev);

	/* Check state and flags */
	if(!(sc->sc_state & HAVEBUS)) {
		LPT_DPRINTF(("%s(%s): attempt to write using device which does "
			"not own the bus(%s).\n", __func__, device_xname(dev),
			device_xname(device_parent(dev))));
		return EINVAL;
	}

	LPT_VPRINTF(("%s(%s): writing %zu bytes\n", __func__,
	    device_xname(dev), uio->uio_resid));

	/* Write the data */
	sc->sc_state &= ~INTERRUPTED;
	while (uio->uio_resid) {
		n = MIN(BUFSIZE, uio->uio_resid);
		error = uiomove(sc->sc_inbuf, n, uio);
		if (error)
			break;

		error = ppbus_write(device_parent(dev), sc->sc_inbuf, n, ioflag,
			&cnt);
		if (error) {
			if (error != EWOULDBLOCK)
				sc->sc_state |= INTERRUPTED;
			break;
		}
	}

	LPT_VPRINTF(("%s(%s): transfer finished, error %d.\n", __func__,
	    device_xname(dev), error));

	return error;
}

/* Printer ioctl */
int
lptioctl(dev_t dev_id, u_long cmd, void *data, int flags, struct lwp *l)
{
	device_t dev = device_lookup(&lpt_cd, LPTUNIT(dev_id));
	struct lpt_softc *sc = device_private(dev);
	int val, fl;
	int error=0;

	if(!(sc->sc_state & HAVEBUS)) {
		LPT_DPRINTF(("%s(%s): attempt to perform ioctl on device which "
			"does not own the bus(%s).\n", __func__, device_xname(dev),
			device_xname(device_parent(dev))));
		return EBUSY;
	}

	switch (cmd) {
	case LPTGMODE:
        	switch (ppbus_get_mode(device_parent(dev))) {
		case PPBUS_COMPATIBLE:
			val = mode_standard;
			break;
		case PPBUS_NIBBLE:
			val = mode_nibble;
			break;
		case PPBUS_PS2:
			val = mode_ps2;
			break;
		case PPBUS_FAST:
			val = mode_fast;
			break;
		case PPBUS_EPP:
			val = mode_epp;
			break;
		case PPBUS_ECP:
			val = mode_ecp;
			break;
		default:
			error = EINVAL;
			val = mode_unknown;
			break;
		}
		*(int *)data = val;
		break;

	case LPTSMODE:
        	switch (*(int *)data) {
		case mode_standard:
			val = PPBUS_COMPATIBLE;
			break;
		case mode_nibble:
			val = PPBUS_NIBBLE;
			break;
		case mode_ps2:
			val = PPBUS_PS2;
			break;
		case mode_fast:
			val = PPBUS_FAST;
			break;
		case mode_epp:
			val = PPBUS_EPP;
			break;
		case mode_ecp:
			val = PPBUS_ECP;
			break;
		default:
			error = EINVAL;
			val = mode_unknown;
			break;
		}

		if (!error)
			error = ppbus_set_mode(device_parent(dev), val, 0);

		break;

	case LPTGFLAGS:
		fl = 0;

		/* DMA */
		error = ppbus_read_ivar(device_parent(dev), PPBUS_IVAR_DMA, &val);
		if (error)
			break;
		if (val)
			fl |= LPT_DMA;

		/* IEEE mode negotiation */
		error = ppbus_read_ivar(device_parent(dev), PPBUS_IVAR_IEEE, &val);
		if (error)
			break;
		if (val)
			fl |= LPT_IEEE;

		/* interrupts */
		error = ppbus_read_ivar(device_parent(dev), PPBUS_IVAR_INTR, &val);
		if (error)
			break;
		if (val)
			fl |= LPT_INTR;

		/* lpt-only flags */
		fl |= sc->sc_flags;

		*(int *)data = fl;
		break;

	case LPTSFLAGS:
		fl = *(int *)data;

		/* DMA */
		val = (fl & LPT_DMA);
		error = ppbus_write_ivar(device_parent(dev), PPBUS_IVAR_DMA, &val);
		if (error)
			break;

		/* IEEE mode negotiation */
		val = (fl & LPT_IEEE);
		error = ppbus_write_ivar(device_parent(dev), PPBUS_IVAR_IEEE, &val);
		if (error)
			break;

		/* interrupts */
		val = (fl & LPT_INTR);
		error = ppbus_write_ivar(device_parent(dev), PPBUS_IVAR_INTR, &val);
		if (error)
			break;

		/* lpt-only flags */
		sc->sc_flags = fl & (LPT_PRIME|LPT_AUTOLF);

		break;

	default:
		error = EINVAL;
		break;
	}

	return error;
}

